### PR TITLE
Add [`manual_slice_size_calculation`]

### PR DESCRIPTION
**Pull request by schubart**
Fixes: #10518

---

changelog: new lint [`manual_slice_size_calculation`]

### CODE REVIEW

File: clippy_lints/src/manual_slice_size_calculation.rs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    }
+}
+
+fn simplify<'tcx>(
+    cx: &LateContext<'tcx>,
+    expr1: &'tcx Expr<'tcx>,
+    expr2: &'tcx Expr<'tcx>,
+) -> Option<&'tcx Expr<'tcx>> {
+    simplify_half(cx, expr1, expr2).or_else(|| simplify_half(cx, expr2, expr1))
+}
```
Comment (llogiq) [inline clippy_lints/src/manual_slice_size_calculation.rs:64]: I swear I've seen this pattern multiple times. Perhaps we should make a generic `clippy_utils` function of it.

**Review by llogiq — COMMENTED** [review]
I'd like to see a test where either the size or length is bound to a temporary variable first. It's ok for now if that isn't linted, but sufficiently easy to change.

Otherwise the code looks good, and I'll gladly merge it with the additional test added.

**Comment by schubart** [comment]
> I'd like to see a test where either the size or length is bound to a temporary variable first. It's ok for now if that isn't linted, but sufficiently easy to change.
> 
> Otherwise the code looks good, and I'll gladly merge it with the additional test added.

Done.

I feel this kind of data flow analysis (where does a value come from, what do we know about it?) could be a whole new thing, applicable to a lot more than just this lint here.

For example, in the following Java code Intellij warns that the `if` condition is always true:

```java
List<String> strings = Collections.emptyList();
if (strings.isEmpty()) {
    System.out.println("There are no strings");
}
```

and in this code it warns that the list access will always fail:

```java
List<String> strings = foo();
if (strings.isEmpty()) {
    System.out.println(strings.get(0));
}
```

Do we have anything like this in Clippy?

**Comment by llogiq** [comment]
We have, it's `clippy_utils::expr_or_init(..)`. If you want to add this directly to this PR, feel free, otherwise I'll have this merged and we can do a follow-up.

**Comment by schubart** [comment]
Thanks, I'll give it a shot. Let's not merge this yet.

**Comment by schubart** [comment]
Done. Thanks for the suggestion, this is very nice!

**Comment by llogiq** [comment]
Great!

@bors r+

**Pull request merged**